# rCom: A route-based framework inferring cell type communication and regulatory network using single cell data
 
## Abstract
### Background
The mapping of ligand-receptor pairs is the cornerstone of understanding complicated intercellular interactions. With recent advances of single cell RNA (scRNA) sequencing technology, several methods have been proposed to infer cell-cell communication by analyzing ligand-receptor pairs. However, existing methods have limited ways of using what we call “prior knowledge”, i.e., what are already known (albeit incompletely) about the upstream for the ligand and the downstream for the receptor.
### Method
In this paper, we present a novel framework, called rCom, capable of inferring cell-cell interactions by considering portions of pathways that would be associated with upstream of the ligand and downstream of receptors under examination. (Fig. 1.) The rCom framework integrates knowledge from multiple biological databases including transcription factor-target database, ligand-receptor database and publicly available curated signaling pathway databases. The rCom framework examines combinatoric ways of integrating the partially known relationships against the cohorts of gene expression datasets obtainable through subtyped cells.  We combine both algorithmic methods and heuristic rules to score how each putative ligand-receptor pair may matchup between all possible cell subtype pairs. Permutation test is performed to rank the hypothesized cell-cell communication routes. 

![image](https://user-images.githubusercontent.com/114254986/215358298-7786e6b6-1ed3-40ab-8e4d-8406fea67752.png)<br>
*Fig 1. The overview architecture of rCom. a. The three categories of biological knowledge databases used in the rCom. b. The input data for rCom including: (i). preprocessed scRNA-seq with real cell group label for inferencing inter cell type communication. (ii). preprocessed scRNA-seq with randomized group label for permutation test. (iii). Cell group label meta data file. c. The matrix of communication score between cell type. d. The communication route pairs extract from 3 type databases. e. Route-wised communication scored in each cell. f. The post analysis of the cell types communication include: inferenced rICRN, summary table including cell type with permutation p-value, violin plot for the ligand-receptor expression value.*

### Result and conclusion

We performed two case studies using single cell transcriptomic data from bone biology. Our literature survey suggests that rCom could be effective in discovering novel cell-cell communication relationships that have been only partially known in the field. (Fig. 2, 3). For details pleae check  <a href="https://dl.acm.org/doi/abs/10.1145/3535508.3545514">paper</a>.

![image](https://user-images.githubusercontent.com/114254986/215358321-e9901a2f-9685-4160-acea-77839c0146c7.png)<br>
*Fig 2.Communication network generated by rCOM using Acta2-lineage cell in osseointegration. The label of edge shows the number of communication routes between two cell groups*<br>
![image](https://user-images.githubusercontent.com/114254986/215358326-0a26a908-acc6-4693-a118-f0da9baf9425.png)<br>
*Fig 3. Communication network generated by rCOM using bone marrow stromal dataset scRNA-seq of 11 weeks old mice dataset*<br>


## Questions
<p><em>For more information please connecting honglin.wang@uconn.edu</em></p>
<p><em>For citation please check the <a href="https://dl.acm.org/doi/abs/10.1145/3535508.3545514">here</a></p>
